### DISCUSSION TITLE
How can a MV3 extension load JS files of other extensions

### DISCUSSION THREAD

**Discussion by gdh1995** [microsoft/MicrosoftEdge-Extensions#68] [Q&A]
I once made a general extension and I want to make it run on all tabs including extension pages (e.g. a PDF viewer extension can support another VIM-like keyboard-shortcut extension), but MV3's `content_security_policy.extension_pages` is so strict that my trick on MV2 breaks. So I wonder whether Edge team can update MV3 policy to re-enable this usage.

Sorry to disturb you, but although I've filed https://bugs.chromium.org/p/chromium/issues/detail?id=1282890 on Dec 27, 2021, there seem no people to take care about it. In fact I haven't seen any outside non-crash issue being solved by the Google Chromium team.

The details of my trick in MV2 is:
1. if only an extension's `manifest.json` declares a speical CSP, like:
- 1. `"content_security_policy": "script-src 'self' chrome-extension://hfjbmagddngcpeloejdejnfgbamkjaeg/; object-src 'self'"`
2. then it's HTML pages will be allowed to run JS files from the extension with the origin
- 1. e.g.: `<script src="chrome-extension://hfjbmagddngcpeloejdejnfgbamkjaeg/lib/injector.js"></script>`
3. and the `lib/injector.js` is also written by myself, and it will send messages, inject other `<script>` elements
4. then the later extension of `hfjbmagddngcpeloejdejnfgbamkjaeg` runs perfectly on the prior extension

My [Vimium C](https://github.com/gdh1995/vimium-c/issues) has been accepted to run on 6 extensions published on Chrome Web Store, and 4 is made by myself while 2 are other extensions which have hundreds of thousands of users (SalaDict and Edge Translate). The version of Vimium C for Edge Add-Ons has 47K+ users and also support this usage.

However, Google's MV3 doesn't allow the `chrome-extension:` protocol in `content_security_policy.extension_pages`, and then my trick breaks. I can not understand why it even allows `http://localhost:<port>` ([as said here](https://developer.chrome.com/docs/extensions/mv3/mv3-migration/#content-security-policy:~:text=disallows%20certain%20CSP%20modifications%20for%20extension_pages)) but insist on refusing other extensions' scripts . I appreciate contributions of you Edge team if anyone can help and add `chrome-extension:` back.

Notes:
1. BTW, Firefox also supports such a trick, and what's more is, because it doesn't need special `CSP` rules, an extension can run scripts from any other extension if only it knows script URLs (there's a random extension UUID in URL generated on first installation).
2. here's a test case: [test-js-injection-in-mv3.zip](https://github.com/microsoft/MicrosoftEdge-Extensions/files/10017823/test-js-injection-in-mv3.zip), and its result snapshot had been published on https://bugs.chromium.org/p/chromium/issues/detail?id=1282890#c3

**Comment by sivsouvamMSFT** [comment]
Hi @gdh1995, thank you for bringing it up. I will check with our Engg team for solutions/possibilities & get back to you.
